### PR TITLE
SYSDM.CPL: Pagefile support improvements

### PR DESCRIPTION
**Pull request by HBelusca**
## Purpose

SYSDM.CPL used hardcoded maximum pagefile size limits in the code.
Now use the architecture-specific limits (see PR #4843 for more details).
Also fix the problem of hardcoded values within the translated strings.

Parts of this PR: addendum to commits 4d2d2dbb2 (PR #2597), 3bee3b92a (PR #2706)

JIRA issues: [CORE-18574](https://jira.reactos.org/browse/CORE-18574), [CORE-12486](https://jira.reactos.org/browse/CORE-12486)

## Proposed changes

Prerequisites:
```
[CPL:SYSDM] Make ResourceMessageBox() support printf-like messages.

+ Make usage of this new functionality.
+ Cleanup precomp.h.
```

And the main commit:
```
[CPL:SYSDM] Improvements for the Paging File settings dialog.

- Implement the architecture-specific pagefile size limits.
  In particular, verify that the selected volume on which to create
  the page file can accomodate the theoretical maximum limit (e.g.
  FAT32 cannot accomodate a 16 TB pagefile on x64, and the limit must
  be lowered down to 4 GB).

- Change the IDS_WARNINITIALRANGE and IDS_WARNMAXIMUMRANGE strings
  so that the maximum limit mentioned is the dynamic one.

- Review, improve and fix other aspects of the code.
```

### CODE REVIEW

**Review by katahiromz — COMMENTED** [review]
I don't know about paging.

File: dll/cpl/sysdm/sysdm.c [hunk 1]
```diff
@@ -23,22 +23,30 @@ APPLET Applets[NUM_APPLETS] =
...
+    LoadStringW(hInstance, uText, szText, _countof(szText));
+
+    va_start(args, uText);
+    StringCchVPrintfW(szCookedText, _countof(szCookedText),
+                      szText, args);
+    va_end(args);
 
     return MessageBoxW(hwnd,
-                       szText,
+                       szCookedText,
```
Comment (binarymaster) [inline dll/cpl/sysdm/sysdm.c:49]: Mmm cooked text, must be delicious :yum:

File: dll/cpl/sysdm/virtmem.c [hunk 2]
```diff
@@ -151,44 +193,48 @@ ParseMemSettings(PVIRTMEM pVirtMem)
...
+                if (PageFile->pszVolume != NULL)
+                    _tcscpy(PageFile->pszVolume, szVolume);
+
+                /*
+                 * Check whether the volume is FAT, which cannot support files
+                 * larger than 4GB (for FAT32 and FATX, and less for FAT16).
+                 * This will limit the maximum size of the pagefile this volume
+                 * can contain (see GetMaxPageFileSize()).
+                 */
+                PageFile->bIsNotFAT = (_tcsnicmp(szFSName, _T("FAT"), 3) != 0);
```
Comment (binarymaster) [inline dll/cpl/sysdm/virtmem.c:237]: I wonder if there's an API to ask filesystem for maximum file size :thinking:

File: dll/cpl/sysdm/lang/ru-RU.rc [hunk 3]
```diff
@@ -356,8 +356,8 @@ BEGIN
     IDS_MESSAGEBOXTITLE "Панель управления системы"
     IDS_WARNINITIALSIZE "Введите числовое значение исходного размера файла подкачки."
     IDS_WARNMAXIMUMSIZE "Введите числовое значение для параметра максимальный размер файла подкачки."
-    IDS_WARNINITIALRANGE "Начальный размер файла подкачки должен быть не меньше, чем 2 МБ и не должен превышать свободного места на выбранном диске."
-    IDS_WARNMAXIMUMRANGE "Максимальный размер файла подкачки должен быть не меньше, чем начальный размер, который не должен превышать свободного места на выбранном диске."
+    IDS_WARNINITIALRANGE "The initial paging file size must be between 2 MB and %ld MB, and must not exceed the available space on the selected drive."
+    IDS_WARNMAXIMUMRANGE "The maximum paging file size must be larger than or equal to its initial size, less than %ld MB, and must not exceed the available space on the selected drive."
```
Comment (binarymaster) [inline dll/cpl/sysdm/lang/ru-RU.rc:360]:
```suggestion
    IDS_WARNINITIALRANGE "Исходный размер файла подкачки должен быть в диапазоне между 2 МБ и %ld МБ, и не должен превышать свободного места на выбранном диске."
    IDS_WARNMAXIMUMRANGE "Максимальный размер файла подкачки должен быть больше или равен своему исходному размеру, при этом меньше %ld МБ, и не должен превышать свободного места на выбранном диске."
```

**Review by binarymaster — COMMENTED** [review]
General suggestion: please use `[SYSDM]` as prefix in commit messages, so it matches CMake module name, not `[SYSDM.CPL]`.

CMake module names are always required to be unique in the source tree, so explicit '.CPL' is not needed.

File: dll/cpl/sysdm/virtmem.c (hunk 2 above)
Comment (HBelusca) [reply dll/cpl/sysdm/virtmem.c:237]: Sadly I've tried to find one but none showed up.

File: dll/cpl/sysdm/lang/de-DE.rc [hunk 4]
```diff
@@ -365,8 +365,8 @@ BEGIN
     IDS_MESSAGEBOXTITLE "Systemsteuerungsoption ""System"""
     IDS_WARNINITIALSIZE "Geben Sie einen numerischen Wert für die Anfangsgröße der Auslagerungsdatei an."
     IDS_WARNMAXIMUMSIZE "Geben Sie einen numerischen Wert für die Maximalgröße der Auslagerungsdatei an."
-    IDS_WARNINITIALRANGE "Die Anfangsgröße der Auslagerungsdatei darf nicht kleiner als 2 MB sein und darf den verfügbaren Speicherplatz auf dem gewählten Laufwerk nicht überschreiten."
-    IDS_WARNMAXIMUMRANGE "Die Maximalgröße der Auslagerungsdatei darf nicht kleiner als die Anfangsgröße sein und darf den verfügbaren Speicherplatz auf dem gewählten Laufwerk nicht überschreiten."
+    IDS_WARNINITIALRANGE "The initial paging file size must be between 2 MB and %ld MB, and must not exceed the available space on the selected drive."
```
Comment (JoachimHenze) [inline dll/cpl/sysdm/lang/de-DE.rc:368]: Please restore the german localization in its updated form, here is it:

IDS_WARNINITIALRANGE "Die Anfangsgröße der Auslagerungsdatei muss zwischen 2 MB und %ld MB sein und darf den verfügbaren Speicherplatz auf dem gewählten Laufwerk nicht überschreiten."
IDS_WARNMAXIMUMRANGE "Die Maximalgröße der Auslagerungsdatei darf nicht kleiner als die Anfangsgröße sein, muss kleiner als %ld MB sein und darf den verfügbaren Speicherplatz auf dem gewählten Laufwerk nicht überschreiten."

**Review by JoachimHenze — CHANGES_REQUESTED** [review]

File: dll/cpl/sysdm/virtmem.c [hunk 5]
```diff
@@ -707,17 +806,18 @@ VirtMemDlgProc(HWND hwndDlg,
...
     {
         case WM_INITDIALOG:
+        {
             pVirtMem = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(VIRTMEM));
             if (pVirtMem == NULL)
             {
                 EndDialog(hwndDlg, 0);
                 return FALSE;
             }
-
```
Comment (JoachimHenze) [inline dll/cpl/sysdm/virtmem.c:716]: Elimination of that line makes no sense.

**Comment by HBelusca** [comment]
Cc @reactos/lang-chinese @reactos/lang-dutch @reactos/lang-estonian @reactos/lang-hebrew @reactos/lang-hindi @reactos/lang-hungarian @reactos/lang-indonesian @reactos/lang-italian @reactos/lang-polish @reactos/lang-portuguese @reactos/lang-romanian @reactos/lang-spanish @reactos/lang-turkish @reactos/lang-ukrainian for translation update :)

**Comment by binarymaster** [comment]
> [CPL:SYSDM]

@HBelusca I would still prefer just `[SYSDM]` for it.

Corner cases are big and important  modules, such as `[NTOS:xxx]` and `[WIN32SS:xxx]`.

**Review by GeoB99 — APPROVED** [review]
No objections from me. I may post the translations both for Italian & Romanian later or I'll just commit such translations myself after merge rather than bothering you.

**Review by JoachimHenze — APPROVED** [review]

**Comment by carlo-bramini** [comment]
Excuse me, I have seen these strings:

```
"The initial paging file size must be between 2 MB and %lu MB, and must not exceed the available space on the selected drive."
"The maximum paging file size must be larger than or equal to its initial size, less than %lu MB, and must not exceed the available space on the selected drive."
```

but it is not very clear to me the reason of having something like that.
Actually, we ***know*** the current amount of available space on the selected volume when these messages come out, so the maximum size could be just clamped to that value and the strings could just print something like this:

```
"The initial paging file size must be between 2 MB and %lu MB on the selected drive."
"The maximum paging file size must be larger than or equal to its initial size and less than %lu MB on the selected drive."
```

Is it correct?

**Comment by HBelusca** [comment]
> Is it correct?

Maybe. Indeed the code now will display the `min(freespace, max_pagefile_size_from_architecture);` (schematically speaking).

**Comment by HBelusca** [comment]
@binarymaster @JoachimHenze @Mas4hmad : translations have been updated again, please recheck :)

**Review by binarymaster — APPROVED** [review]
I would still prefer `[SYSDM]` prefix instead of `[CPL:SYSDM]`. Translation is correct. :+1:

**Comment by carlo-bramini** [comment]
Italian translation:

IDS_WARNINITIALRANGE:
"The initial paging file size must be between 2 MB and %lu MB on the selected drive."
"La dimensione iniziale del file di paging non può essere inferiore a 2 MB e %lu MB nell'unità selezionata."

IDS_WARNMAXIMUMRANGE:
"The maximum paging file size must be larger than or equal to its initial size, and less than %lu MB on the selected drive."
"La dimensione massima del file di paging deve essere superiore a quella iniziale e minore di %lu MB nell'unità selezionata."

**Comment by HBelusca** [comment]
@carlo-bramini Thanks for your translation, and sorry for the 2-year long delay. See PR #7216 for the corresponding PR.

**Pull request merged**